src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -233,29 +234,27 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
                 return;
             }
 
-            object? propertyValue = property.GetValue(instance);
-            bool hasSetter = property.SetMethod != null && (property.SetMethod.IsPublic || options.BindNonPublicProperties);
-
-            if (propertyValue == null && !hasSetter)
-            {
-                // Property doesn't have a value and we cannot set it so there is no
-                // point in going further down the graph
-                return;
-            }
+            var propertyBindingPoint = new BindingPoint(
+                initialValueProvider: () => property.GetValue(instance),
+                isReadOnly: property.SetMethod is null || (!property.SetMethod.IsPublic && !options.BindNonPublicProperties));
 
-            propertyValue = GetPropertyValue(property, instance, config, options);
+            BindInstance(
+                property.PropertyType,
+                propertyBindingPoint,
+                config.GetSection(GetPropertyName(property)),
+                options);
 
-            if (propertyValue != null && hasSetter)
+            if (propertyBindingPoint.HasNewValue)
             {
-                property.SetValue(instance, propertyValue);
+                property.SetValue(instance, propertyBindingPoint.Value);
             }
         }
 
         [RequiresUnreferencedCode("Cannot statically analyze what the element type is of the object collection in type so its members may be trimmed.")]
-        private static object? BindToCollection(Type type, IConfiguration config, BinderOptions options)
+        private static object BindToCollection(Type type, IConfiguration config, BinderOptions options)
         {
             Type genericType = typeof(List<>).MakeGenericType(type.GenericTypeArguments[0]);
-            object? instance = Activator.CreateInstance(genericType);
+            object instance = Activator.CreateInstance(genericType)!;
```

Comment:
Calling Activator.CreateInstance on a List can't ever return null, right?

Reply:
It is unlikely to return null for List<T>

Reply:
That's what I thought; just wanted to highlight the change.